src/Logary.Facade/Facade.fs

```diff
@@ -226,27 +226,194 @@ type LoggingConfig =
+              | LevelInfo -> ConsoleColor.White
+              | LevelWarning -> ConsoleColor.Yellow
+              | LevelError -> ConsoleColor.Red
+              | LevelFatal -> ConsoleColor.Red
+              | KeywordSymbol -> ConsoleColor.Blue
+              | NumericSymbol -> ConsoleColor.Magenta
+              | StringSymbol -> ConsoleColor.Cyan
+              | OtherSymbol -> ConsoleColor.Green
+              | NameSymbol -> ConsoleColor.Gray
+      PrintTemplateFieldNames = false }
+  static member CreateInvariant() =
+    LiterateContext.Create(Globalization.CultureInfo.InvariantCulture)
+
+module internal FsMtParser =
+  open System.Text
+
+  type Property(name:string, format: string) =
+      static let emptyInstance = Property("", null)
+      static member Empty = emptyInstance
+      member __.Name = name
+      member __.Format = format
+      member internal x.AppendPropertyString(sb: StringBuilder, ?replacementName) =
+          let replacementName = defaultArg replacementName name
+          sb.Append("{").Append(replacementName)
+            .Append(match x.Format with null | "" -> "" | _ -> ":" + x.Format).Append("}")
+      override x.ToString() = x.AppendPropertyString(StringBuilder()).ToString()
+
+  module internal Internal =
+
+    let inline isLetterOrDigit c = System.Char.IsLetterOrDigit c
+    let inline isValidInPropName c = c = '_' || System.Char.IsLetterOrDigit c
+    let inline isValidInFormat c = c <> '}' && (c = ' ' || isLetterOrDigit c || System.Char.IsPunctuation c)
+    let inline isValidCharInPropTag c = c = ':' || isValidInPropName c || isValidInFormat c
+
+    [<Struct>]
+    type Range(startIndex:int, endIndex:int) =
+        member inline this.Start = startIndex
+        member inline this.End = endIndex
+        member inline this.Length = (endIndex - startIndex) + 1
+        member inline this.GetSubString (s:string) = s.Substring(startIndex, this.Length)
+        member inline this.IsEmpty = startIndex = -1 && endIndex = -1
+        static member Substring (s:string, startIndex, endIndex) = s.Substring(startIndex, (endIndex - startIndex) + 1)
+        static member Empty = Range(-1, -1)
+
+    let inline tryGetFirstCharInRange predicate (s:string) (range:Range) =
+        let rec go i =
+            if i > range.End then -1
+            else if not (predicate s.[i]) then go (i+1) else i
+        go range.Start
+
+    let inline tryGetFirstChar predicate (s:string) first =
+        tryGetFirstCharInRange predicate s (Range(first, s.Length - 1))
+
+    let inline hasAnyInRange predicate (s:string) (range:Range) =
+        match tryGetFirstChar (predicate) s range.Start with
+        | -1 -> false | i -> i <= range.End
+
+    let inline hasAny predicate (s:string) =
+        hasAnyInRange predicate s (Range(0, s.Length - 1))
+
+    let inline indexOfInRange s range c = tryGetFirstCharInRange ((=) c) s range
+
+    let inline tryGetPropInRange (template:string) (within : Range) : Property =
```

Comment:
This is a log of code that's pretty opaque. Could you explain using some code comments how it's put together?